docs/dev_guides/new_library_comparisons_and_levels.md

```diff
@@ -0,0 +1,286 @@
+
+To create a new comparison level, you must create a new subclass of `ComparisonLevel` which will serve as the base comparison level for any SQL dialects that will allow this level, in the file [`splink/comparison_level_library.py`](https://github.com/moj-analytical-services/splink/blob/master/splink/comparison_level_library.py).
+It will contain the full logic for creating the comparison level - any dialect dependencies will be implemented as properties on the specific dialect-dependent object.
+In this case we will need to refer to a property `_array_length_function_name`, as this can vary by dialect.
+We will not define the property directly on this object, as our dialect-dependent versions will inherit this property from elsewhere.
+We also include any customisable parameters for our level - in this case we will allow options for the maximum number of elements the array may differ by.
+
+```python
+class ArrayLengthLevelBase(ComparisonLevel):
+    def __init__(
+        self,
+        col_name: str,
+        length_difference: int,
+        m_probability=None,
+    ):
+        """Compares two arrays whose sizes are within a fixed distance
+        | length(arr_1) - length(arr_2) | <= (length_difference)
+
+        Arguments:
+            col_name (str): Input column name
+            length_difference (int): Maximum difference in array lengths
+            m_probability (float, optional): Starting value for m probability.
+                Defaults to None.
+
+        Returns:
+            ComparisonLevel: A comparison level that evaluates the size difference
+                between two arrays
+        """
+        col = InputColumn(col_name, sql_dialect=self._sql_dialect)
+        col_l, col_r = col.names_l_r()
+
+        sql_exp = (
+            f"abs("
+            f"{self._array_length_function_name}({col_l}) - "
+            f"{self._array_length_function_name}({col_r})"
+            f") <= {length_difference}"
+        )
+        level_dict = {
+            "sql_condition": sql_exp,
+            "label_for_charts": f"Array sizes differ by at most {length_difference)",
+        }
+
+        if m_probability:
+            level_dict["m_probability"] = m_probability
+
+        super().__init__(level_dict, sql_dialect=self._sql_dialect)
+```
+
+If you are using a new dialect-dependent property (as we are in this case), then it should be added as a property on `DialectBase` in [`splink.dialect_base.py`](https://github.com/moj-analytical-services/splink/blob/master/splink/dialect_base.py),
+with either a sensible default value, or raising a `NotImplementedError`:
+
+```python
+class DialectBase():
+    ...
+    @property
+    def _array_length_function_name():
+        return "array_length"
```

Comment:
I think this should include a `NotImplementedError`, instead of returning a value, so it better aligns with [your other tweaks](https://github.com/moj-analytical-services/splink/blob/4e7d654b76c93136b3f31422bf8cf2d0e123f41a/splink/dialect_base.py#L13).

Reply:
Or perhaps you should keep add it as an example?

I just feel that users should be pushed to implement a not implemented error, rather than a default value.